proxy/context.go

```diff
@@ -0,0 +1,170 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/zalando/skipper/routing"
+)
+
+type context struct {
+	responseWriter        http.ResponseWriter
+	request               *http.Request
+	response              *http.Response
+	route                 *routing.Route
+	deprecatedServed      bool
+	servedWithResponse    bool // to support the deprecated way independently
+	pathParams            map[string]string
+	stateBag              map[string]interface{}
+	originalRequest       *http.Request
+	originalResponse      *http.Response
+	outgoingHost          string
+	debugFilterPanics     []interface{}
+	outgoingDebugRequest  *http.Request
+	incomingDebugResponse *http.Response
+	startServe            time.Time
+}
+
+func defaultBody() io.ReadCloser {
+	return ioutil.NopCloser(&bytes.Buffer{})
+}
+
+func defaultResponse(r *http.Request) *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusNotFound,
+		Header:     make(http.Header),
+		Body:       defaultBody(),
+		Request:    r,
+	}
+}
+
+func cloneURL(u *url.URL) *url.URL {
+	uc := *u
+	return &uc
+}
+
+func cloneRequestMetadata(r *http.Request) *http.Request {
```

Comment:
Why is clone useful ?

This is in context of `r.Body` can be read only once and you want to consume it multiple times. But here, since the body is always replaced with `defaultBody()`, why is cloning the metadata useful ? 

Also, same question for `cloneResponseMetadata`.